sdk_extensions/logging/support/src/main/java/io/opentelemetry/logging/BatchLogProcessor.java

```diff
@@ -0,0 +1,372 @@
+    }
+
+    private static final BoundLongCounter exporterFailureCounter;
+    private static final BoundLongCounter exporterBusyCounter;
+    private static final BoundLongCounter droppedRecordCounter;
+    private static final Logger logger = Logger.getLogger(Worker.class.getName());
+
+    private final Object monitor = new Object();
+    private final int maxQueueSize;
+    private final long scheduleDelayMillis;
+    private final ArrayList<LogRecord> logRecords;
+    private final int maxExportBatchSize;
+    private final ExecutorService executor;
+    private final Exporter exporter;
+    private final Timer timer = new Timer(TIMER_THREAD_NAME, /* isDaemon= */ true);
+    private final long exporterTimeout;
+
+    private Worker(
+        int maxQueueSize,
+        long scheduleDelayMillis,
+        int maxExportBatchSize,
+        long exporterTimeout,
+        Exporter exporter) {
+      this.maxQueueSize = maxQueueSize;
+      this.maxExportBatchSize = maxExportBatchSize;
+
+      this.exporterTimeout = exporterTimeout;
+      this.scheduleDelayMillis = scheduleDelayMillis;
+      this.exporter = exporter;
+      this.logRecords = new ArrayList<>(maxQueueSize);
+      // We should be able to drain a full queue without dropping a batch
+      int exportQueueSize = maxQueueSize / maxExportBatchSize;
+      RejectedExecutionHandler h =
+          new RejectedExecutionHandler() {
+            @Override
+            public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+              exporterBusyCounter.add(1);
+            }
+          };
+      this.executor =
+          new ThreadPoolExecutor(
+              1,
+              1,
+              0L,
+              TimeUnit.MILLISECONDS,
+              new LinkedBlockingQueue<Runnable>(exportQueueSize),
+              h);
+    }
+
+    @Override
+    public void run() {
+      ArrayList<LogRecord> logsCopy;
+      while (!Thread.currentThread().isInterrupted()) {
+        synchronized (monitor) {
+          if (this.logRecords.size() < maxExportBatchSize) {
+            do {
+              try {
+                monitor.wait(scheduleDelayMillis);
+              } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                return;
+              }
+            } while (this.logRecords.isEmpty());
+          }
+          logsCopy = new ArrayList<>(this.logRecords);
+          logRecords.clear();
+        }
+        exportBatches(logsCopy);
+      }
+    }
+
+    private void exportBatches(ArrayList<LogRecord> recordsToShip) {
+      for (int i = 0; i < recordsToShip.size(); ) {
+        int lastIndexToTake = Math.min(i + maxExportBatchSize, recordsToShip.size());
+        onBatchExport(createLogDataForExport(recordsToShip, i, lastIndexToTake));
+        i = lastIndexToTake;
+      }
+    }
+
+    private void onBatchExport(final List<LogRecord> logDataForExport) {
+      final Future<CompletableResultCode> f =
```

Comment:
This was a bit that gave me fits. In the span export, there's a monitor that dumps a batch if a batch is already in process. Unfortunately exportBatches is set up to export several batches at the same time, meaning that the first will lock subsequent batches, so if the incoming queue fills, the first batch goes to the exporter and the rest are silently discarded. The intent here is to use an execution queue to allow enough exports to occur to drain a full input queue, but still retain the behavior of shedding batches if the exporter can't keep up.